src/update/update.cpp

```diff
@@ -26,505 +26,496 @@
-                        update->Name().c_str(), pIndexPrev->nHeight, minerUpdateResult.ToString().c_str(), currentThreshold);
-                  }
-               }
-
-               // Check Node votes:
-               bool nodesApprove = false;
-               {
-                  VoteResult nodeUpdateResult = nodeUpdateVoting.GetVote(pIndexPrev, *update);
-
-                  int64_t nodeMean = nodeUpdateResult.MeanPercent();
-                  int64_t currentThreshold = update->NodeThreshold().GetThreshold(roundNumber);
-                  if (nodeMean >= currentThreshold)
-                  {
-                     LogPrint(BCLog::UPDATES, "Updates: Update: %s, Voting: Height: %7d, Nodes  %s, Threshold: %3d - Nodes  approve\n",
-                        update->Name().c_str(), pIndexPrev->nHeight, nodeUpdateResult.ToString().c_str(), currentThreshold);
-                     nodesApprove = true;
-                  }
-                  else
-                  {
-                     LogPrint(BCLog::UPDATES, "Updates: Update: %s, Voting: Height: %7d, Nodes  %s, Threshold: %3d\n",
-                        update->Name().c_str(), pIndexPrev->nHeight, nodeUpdateResult.ToString().c_str(), currentThreshold);
-                  }
-               }
-
-               if (minersApprove && nodesApprove)
-               {
-                  stateNext = { EUpdateState::LockedIn, pIndexPrev->nHeight + roundSize * update->GraceRounds() };
-                  LogPrint(BCLog::UPDATES, "Updates: Update: %s, Voting: Height: %7d, Proposal has been locked in, Activation at height: %7d\n", update->Name().c_str(), pIndexPrev->nHeight, stateNext.FinalHeight);
-               }
-               break;
+            case EUpdateState::Voting: {
+                if (pIndexPrev->nHeight > lastVotingHeight) {
+                    if (update->ForcedUpdate()) {
+                        stateNext = {EUpdateState::LockedIn, pIndexPrev->nHeight + roundSize * update->GraceRounds(),
+                                     voteStats};
+                        LogPrint(BCLog::UPDATES,
+                                 "Updates: Update: %s, Voting: Height: %7d, Voting failed, forcing update, Activation at height: %7d\n",
+                                 update->Name().c_str(), pIndexPrev->nHeight, stateNext.FinalHeight);
+                    } else {
+                        stateNext = {EUpdateState::Failed, pIndexPrev->nHeight, stateInfo.voteStats};
+                    }
+                    break;
+                }
+
+                // Check Miner votes:
+                {
+                    VoteResult minerUpdateResult = minerUpdateVoting.GetVote(pIndexPrev, *update);
+                    if (voteStats.minerVoteResult) {
+                        free(voteStats.minerVoteResult);
+                    }
+                    voteStats.minerVoteResult = &minerUpdateResult;
```

Comment:
You can't keep an address of a variable on the stack (and potentially later call free on it).